src/metabase/models/card.clj

```diff
@@ -767,6 +767,96 @@
+  [breakout-clause]
+  (reduce (fn [acc breakout]
+            (update acc (subvec breakout 0 2)
+                    #(conj (set %1) breakout)))
+          {}
+          breakout-clause))
+
+(defn- action-for-identifier+refs
+  "Generate _action_ for combination of args.
+
+  _Action_ is to be performed on _parameter mapping_ of a _dashcard_. For more info see
+  the [[update-associated-parameters!]]'s docstring.
+
+  _Action_ has a form of [<action> & args]. Currently :update, :noop are implemented."
+  [after--identifier->refs identifier before--refs]
+  (let [after--refs (get after--identifier->refs identifier #{})]
+    (if (and (= 1 (count before--refs) (count after--refs))
+             (not= before--refs after--refs))
+      [:update (first after--refs)]
+      [:noop])))
+
+(defn- breakouts-->identifier->action
+  "Generate mapping of _identifier_ -> _action_.
+
+  _identifier_ is is a vector of first 2 elements of ref, eg. [:expression \"xix\"] or [:field 10]. Action is generated
+  in [[action-for-identifier+refs]] and performed later in [[update-mapping]]."
+  [breakout-before-update breakout-after-update]
+  (let [before--identifier->refs (breakout-->identifier->refs breakout-before-update)
+        after--identifier->refs  (breakout-->identifier->refs breakout-after-update)
+        action (partial action-for-identifier+refs after--identifier->refs)
+        action-kvrf (fn [acc identifier before--refs] (assoc acc identifier (action identifier before--refs)))]
+    ;; Remove no-ops to avoid redundant db calls in [[update-associated-parameters!]].
+    (not-empty (m/filter-vals (complement #{[:noop]}) (reduce-kv action-kvrf {} before--identifier->refs)))))
+
+(defn- update-mapping
+  "Return modifed mapping, or nil, according to action."
+  [identifier->action mapping]
+  (if-not (and (= :dimension (get-in mapping [:target 0]))
+               (#{:field :expression} (get-in mapping [:target 1 0])))
+    mapping
+    (let [dimension (get-in mapping [:target 1])
+          identifier (subvec dimension 0 2)
+          [action arg] (get identifier->action identifier [:noop])]
+      (case action
+        :update (assoc-in mapping [:target 1] arg)
+        :noop   mapping))))
+
+(defn- update-for-dashcard
+  "Return nil for no-op updates or [<dashcard-id> <update>]. <update> is a map of `:parameter_mappings` only as only
+  this is updated."
+  [identifier->action dashcard]
+  (let [card-update (select-keys (update dashcard :parameter_mappings
+                                         (comp vec (partial keep (partial update-mapping identifier->action))))
+                                 [:parameter_mappings])]
+    (when (not= (:parameter_mappings dashcard) (:parameter_mappings card-update))
+      [(:id dashcard) card-update])))
+
+(defn- updates-for-dashcards
+  [identifier->action dashcards]
+  (not-empty (keep (partial update-for-dashcard identifier->action) dashcards)))
+
+(defn- update-associated-parameters!
```

Comment:
Just calling it out for future reference: all of this becomes dead code with the refs overhaul. Parameters mappings would use the new `:ident`s, which are consistently applied to a particular breakout. Editing its temporal unit (either directly or via a parameter) won't change the `:ident`, and so won't change the parameter mapping.

I've got a big "hit list" for cleanup after refs are landed; this is on it.

Reply:
> after refs are landed

That is going to be a big day for the qp :)